### PR TITLE
Add conversion webhook for crd apiversion migration

### PR DESCRIPTION
**Pull request by tylerphelan**
- Only used for images and builds which we know will require conversion
- Refactor reusable api types to pkg/apis/core

fix https://github.com/pivotal/kpack/issues/766

Co-authored-by: Tom Kennedy <ktom@vmware.com>

### CODE REVIEW

File: pkg/apis/build/v1alpha2/build_conversion.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"context"
+
+	"knative.dev/pkg/apis"
+
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+)
+
+func (b *Build) ConvertTo(_ context.Context, to apis.Convertible) error {
+	toBuild := to.(*v1alpha1.Build)
```
Comment (tomkennedy513) [inline pkg/apis/build/v1alpha2/build_conversion.go:12]: i think in the future we should look at doing something like this with the type switch https://github.com/tektoncd/pipeline/blob/67b318ba62d9dfee4e9ec639571754f42d9702d7/pkg/apis/pipeline/v1alpha1/cluster_task_conversion.go#L31-L50

**Review by tomkennedy513 — APPROVED** [review]

File: pkg/apis/build/v1alpha2/build_conversion.go (hunk 1 above)
Comment (tylerphelan) [reply pkg/apis/build/v1alpha2/build_conversion.go]: yea we should probably have this, updated

**Review by tomkennedy513 — APPROVED** [review]

**Review by dumez-k — APPROVED** [review]
Looks great! 

Talking with @tylerphelan a little bit and ideally somewhere down the line it might be nice to refactor `corev1alpha1` into it's own package just called `core` rather than versioning it as I find that it leads to confusion a little bit between the two api versions. 

I recognize though that this would be a fairly expensive refactor for the value-add. Just a call-out.

File: pkg/apis/build/v1alpha1/image_types.go [hunk 2]
```diff
@@ -105,4 +96,12 @@ func (i *Image) NamespacedName() types.NamespacedName {
 	return types.NamespacedName{Namespace: i.Namespace, Name: i.Name}
 }
 
+func (i *Image) ConvertTo(_ context.Context, _ apis.Convertible) error {
+	return errors.New("called convertTo in v1alpha1")
```
Comment (matthewmcnew) [inline pkg/apis/build/v1alpha1/image_types.go:100]: v1alpaha2 is the hub?

File: cmd/webhook/main.go [hunk 3]
```diff
@@ -90,6 +93,37 @@ func validatingAdmissionController(ctx context.Context, _ configmap.Watcher) *co
...
 }
 
+func conversionController(ctx context.Context, _ configmap.Watcher) *controller.Impl {
+
+	conversions := map[schema.GroupKind]conversion.GroupKindConversion{
+		buildapi.Kind("Image"): {
+			DefinitionName: "images.kpack.io",
+			HubVersion:     buildapi.SchemeGroupVersion.Version,
+			Zygotes: map[string]conversion.ConvertibleObject{
+				buildapi.SchemeGroupVersion.Version: &buildapi.Image{},
```
Comment (tylerphelan) [inline cmd/webhook/main.go:103]: use `v1alpha2` to be more explicit about conversions

File: pkg/apis/build/v1alpha1/image_types.go (hunk 2 above)
Comment (tylerphelan) [reply pkg/apis/build/v1alpha1/image_types.go:100]: update error message for more context

**Review by matthewmcnew — APPROVED** [review]
Excellent.

**Pull request merged**